Cloudflare: refactor DNS01 challenge to use API for finding the nearest Zone

thiscantbeserious opened the pull request:
**What this PR does / why we need it**:

Bottomline the current implementation isn't using the official Cloudflare API for finding the nearest Zone-ID, instead it uses the generic utility method `/pkg/issuer/acme/dns/util/wait.go#FindZoneByFqdn` which tries to find the Zone by getting SOA records for the public domain. If there are any DNS-Issues (for example if your local DNS-Resolver e.g. is misconfigured or even more common - the Domain is simply incorrectly setup) and there are no SOA-Records returned then this method does not work at all. Or also common in relation with the DNS-Challenge if the Domain in question points to a LAN-Address if its just used for certification.

Furthermore FindZoneByFqdn also is bugged in that it returns the TLD as the default zone to try to access the Cloudflare API in case there are no SOA-Records found at all - e.g. Zone "com" - in general for cloudflare this should be the SLD in combination with the TLD as the default (and likely for others too). Since I do know to little about the specific bits of this generic utility code I didn't touch that part yet.

Instead what I did is refactor the code in the DNS-Issuer for Cloudflare to use the official API for querying the DNS-Tree in the same way, just trought the API (and also fixing that it won't traverse the TLD as a possible Zone-Entry-Name in that regard, like mentioned above).


I've added mocks and a basic unit-test that doesn't need e2e testing to validate the basic-functionality described in the comments.

This was also tested with my own domain in my local test-environment (setup via official means from /devel/README.md)

**Which issue this PR fixes**: 

fixes #4134

**Special notes for your reviewer**:

- See my comments in regard to solving the underlying issue and to get a better idea of the issue itself.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Cloudflare: Refactored DNS01 challenge to use API for finding the nearest Zone (fixing potential DNS-Issues)
```

thiscantbeserious commented on pkg/issuer/acme/dns/cloudflare/cloudflare.go:
```diff
@@ -76,6 +89,46 @@ func NewDNSProviderCredentials(email, key, token string, dns01Nameservers []stri
+// This will try to traverse the official Cloudflare API to find the nearest valid Zone.
+// It's a replacement for /pkg/issuer/acme/dns/util/wait.go#FindZoneByFqdn
+//  example.com.                                   ← Zone-Record found for the SLD (in most cases)
+//  └── foo.example.com.						   ← Zone-Record could be possibly here, but in this case not.
```
Maybe I should fix indentation here - not using tabstops ...

thiscantbeserious commented:
/cc munnerz
/meow

Sorry if I'm disturbing you - its actually my first PR here and I'm unsure about the exact process

irbekrm commented:
/ok-to-test

irbekrm commented:
Hi @thiscantbeserious thank you for the PR!

Could you please try to run `./hack/update-all.sh` - it looks like some deps may be out of date so the tests are failing.

> Sorry if I'm disturbing you - its actually my first PR here and I'm unsure about the exact process

Sorry for the delay! Feel free to chat to us on [#cert-manager-dev](https://kubernetes.slack.com/) about any dev-related things 😄

thiscantbeserious commented:
> Hi @thiscantbeserious thank you for the PR!
> 
> Could you please try to run `./hack/update-all.sh` - it looks like some deps may be out of date so the tests are failing.

Done that, only thing was missing is seemingly a License File from the command above.

pkg/issuer/acme/dns/cloudflare/BUILD.bazel was updated already (maybe I forgot to add it beforehand)

I squashed all commits and removed the tabstops ... lets see ...

thiscantbeserious commented:
@irbekrm not sure if that's due to my changes but now there seems to be a different error ...
````
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //test/integration/acme:go_default_test
-----------------------------------------------------------------------------
--- FAIL: TestAcmeOrdersController (23.13s)
    apiserver.go:64: failed to start control plane: unable to start control plane itself: failed to start the controlplane. retried 5 times: timeout waiting for process etcd to start successfully (it may have failed to start, or stopped unexpectedly before becoming ready)
````

thiscantbeserious commented:
@irbekrm seems fixed now, license file was missing 😄

jakexks commented on pkg/issuer/acme/dns/cloudflare/cloudflare_test.go:
```diff
@@ -9,12 +9,14 @@ this directory.
 package cloudflare
 
 import (
+	"encoding/json"
```
nit: our [code conventions](https://cert-manager.io/docs/contributing/coding-conventions/) for import formatting state this should be split into 3 blocks

```
standard library

external imports

internal imports
```

thiscantbeserious replied:
done :)

jakexks reviewed (commented):
Thanks for this change, looks good. Just one trivial formatting change and I'll be happy to merge it :)

jakexks commented:
/kind bug

thiscantbeserious commented:
... pew almost made my heart stand still god damn ./hack/update-all.sh ... followed by an incorrect squash onto wrong branch 🍔

jakexks commented:
/test pull-cert-manager-bazel

thiscantbeserious commented:
@jakexks tests done :)

jakexks reviewed (approved):
/lgtm
